overlord/devicestate/devicemgr.go

```diff
@@ -1933,18 +1894,70 @@ func (m *DeviceManager) SystemAndGadgetAndEncryptionInfo(wantedSystemLabel strin
 		return nil, nil, nil, fmt.Errorf("cannot parse gadget.yaml: %v", err)
 	}
 
-	encInfo, err := m.encryptionSupportInfo(sys.Model, kernelInfo, gadgetInfo)
+	// Encryption details
+	encInfo, err := m.encryptionSupportInfo(sys.Model, snapInfos[snap.TypeKernel], gadgetInfo)
 	if err != nil {
 		return nil, nil, nil, err
 	}
+
+	// Make sure gadget is valid for model and available encryption
 	opts := &gadget.ValidationConstraints{
 		EncryptedData: encInfo.StorageSafety == asserts.StorageSafetyEncrypted,
 	}
 	if err := gadget.Validate(gadgetInfo, sys.Model, opts); err != nil {
 		return nil, nil, nil, fmt.Errorf("cannot validate gadget.yaml: %v", err)
 	}
 
-	return sys, gadgetInfo, &encInfo, nil
+	return sys, gadgetInfo, &encInfo, err
+}
+
+// loadSystemAndEssentialSnaps loads information for the given label, which
+// includes system, gadget information, gadget and kernel snaps info,
+// and gadget and kernel seed snap info.
+func (m *DeviceManager) loadSystemAndEssentialSnaps(wantedSystemLabel string, types []snap.Type) (
+	*System, map[snap.Type]*snap.Info, map[snap.Type]*seed.Snap, error) {
+
+	// get current system as input for loadSeedAndSystem()
+	systemMode := m.SystemMode(SysAny)
+	currentSys, _ := currentSystemForMode(m.state, systemMode)
+
+	s, sys, err := loadSeedAndSystem(wantedSystemLabel, currentSys)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	// 2. get the gadget volumes for the given system-label
+	perf := &timings.Timings{}
+	if err := s.LoadEssentialMeta(types, perf); err != nil {
+		return nil, nil, nil, fmt.Errorf("cannot load essential snaps metadata: %v", err)
+	}
+	// EssentialSnaps is always ordered, see asserts.Model.EssentialSnaps:
+	// "snapd, kernel, boot base, gadget." and snaps not loaded above
+	// like "snapd" will be skipped and not part of the EssentialSnaps list
+	//
+	snapInfos := make(map[snap.Type]*snap.Info)
+	seedSnaps := make(map[snap.Type]*seed.Snap)
+	for i, typ := range types {
+		seedSnap := s.EssentialSnaps()[i]
```

Comment:
types could be out of order with respect of EssentialSnap, it shouldn't matter. This can do:

```
for _, seedSnap := range s.EssentialSnaps()
```

Reply:
Done